src/cache.c

```diff
@@ -107,102 +115,158 @@ static inline void hlist_del_init(struct hlist_node *n)
-            goto free_lists;
-        INIT_LIST_HEAD(cache->lists[i]);
-    }
 
-    cache->map = malloc(sizeof(hashtable_t));
-    if (!cache->map)
-        goto free_lists;
+    INIT_LIST_HEAD(&cache->list);
+    cache->size = 0;
+    cache->capacity = cache_size;
 
-    cache->map->ht_list_head = malloc(cache_size * sizeof(struct hlist_head));
-    if (!cache->map->ht_list_head)
-        goto free_map;
+    /* prevent overflowing on unsigned integer subtraction */
+    cache->counter = N_CACHE_HISTORY;
 
-    for (size_t ii = 0; ii < cache_size; ii++)
-        INIT_HLIST_HEAD(&cache->map->ht_list_head[ii]);
-    cache->list_size = 0;
-    cache_mp =
-        mpool_create(cache_size * sizeof(lfu_entry_t), sizeof(lfu_entry_t));
-    cache->capacity = cache_size;
-    return cache;
+    cache->map.ht_list_head = malloc(cache_size * sizeof(struct hlist_head));
+    if (!cache->map.ht_list_head) {
+        free(cache);
+        return NULL;
+    }
 
-free_map:
-    free(cache->map);
-free_lists:
-    for (int j = 0; j < i; j++)
-        free(cache->lists[j]);
+    for (uint32_t i = 0; i < cache_size; i++)
+        INIT_HLIST_HEAD(&cache->map.ht_list_head[i]);
 
-    free(cache);
-    return NULL;
+    return cache;
 }
 
 void *cache_get(const cache_t *cache, uint32_t key, bool update)
 {
-    if (!cache->capacity ||
-        hlist_empty(&cache->map->ht_list_head[cache_hash(key)]))
+    if (unlikely(!cache->capacity))
         return NULL;
 
-    lfu_entry_t *entry = NULL;
+    if (hlist_empty(&cache->map.ht_list_head[cache_hash(key)]))
+        return NULL;
+
+    cache_entry_t *entry = NULL;
 #ifdef __HAVE_TYPEOF
-    hlist_for_each_entry (entry, &cache->map->ht_list_head[cache_hash(key)],
+    hlist_for_each_entry (entry, &cache->map.ht_list_head[cache_hash(key)],
                           ht_list)
 #else
-    hlist_for_each_entry (entry, &cache->map->ht_list_head[cache_hash(key)],
-                          ht_list, lfu_entry_t)
+    hlist_for_each_entry (entry, &cache->map.ht_list_head[cache_hash(key)],
+                          ht_list, cache_entry_t)
 #endif
     {
         if (entry->key == key)
             break;
     }
-    if (!entry || entry->key != key)
+
+    if (!entry || entry->key != key || !entry->alive)
         return NULL;
 
+    /*
+     * FIXME: In system simulation, there might be several identical PC from
+     * different processes. We need to check the SATP CSR to update the correct
+     * entry.
+     */
     /* When the frequency of use for a specific block exceeds the predetermined
      * THRESHOLD, the block is dispatched to the code generator to generate C
      * code. The generated C code is then compiled into machine code by the
      * target compiler.
      */
-    if (update && entry->frequency < THRESHOLD) {
-        list_del_init(&entry->list);
-        list_add(&entry->list, cache->lists[entry->frequency++]);
+    if (update) {
+        entry->last_used = cache->counter;
+        entry->frequency++;
```

Comment:
> Would it be safe to assume the frequency number never overflows?

No, however, since the frequency is greater than `THRESHOLD`, the tiered compilation is triggered. After the blocks are compiled and marked as `hot`, the value of frequency has no other effect, even it might overflow from `UINT32_MAX`.